core/src/main/java/org/infinispan/interceptors/distribution/TxDistributionInterceptor.java

```diff
@@ -631,10 +636,15 @@ protected void handleRemotelyRetrievedKeys(InvocationContext ctx, List<?> remote
       assert !mutationsIterator.hasNext();
    }
 
-   private static List<Mutation> getMutationsOnKey(TxInvocationContext ctx, Object key) {
+   private static List<Mutation> getMutationsOnKey(TxInvocationContext ctx, WriteCommand untilCommand, Object key) {
       List<Mutation> mutations = new ArrayList<>();
       // We don't use getAllModifications() because this goes remote and local mods should not affect it
       for (WriteCommand write : ctx.getCacheTransaction().getModifications()) {
+         if (write == untilCommand) {
+            // We've reached this command in the modifications list; this happens when we're replaying a prepared
+            // transaction - see EntryWrappingInterceptor.wrapEntriesForPrepareAndApply
```

Comment:
I'm not sure this is relevant, but is it really true? To me it seems like we always call `getMutationsOnKey(ctx, command, key)` on the originator when we don't have the entry in the context.